groups/bsl/bslstl/bslstl_variant.h

```diff
@@ -0,0 +1,3338 @@
+// alternative.
+template <class TYPE, class VARIANT>
+struct Variant_HasUniqueType;  // Primary template not defined.
+
+template <class TYPE, class HEAD, class... TAIL>
+struct Variant_HasUniqueType<TYPE, bsl::variant<HEAD, TAIL...> >
+: bsl::integral_constant<bool,
+                         Variant_CountType<TYPE, HEAD, TAIL...>::value == 1> {
+};
+#endif
+
+                           // ======================
+                           // struct Variant_ImpUtil
+                           // ======================
+
+template <class VARIANT>
+struct Variant_ImpUtil {
+    // Placeholder class for all functions that need to have friendship access
+    // to the bsl::variant class
+    template <class RET, size_t INDEX>
+    static RET& getImpl(bsl::in_place_index_t<INDEX>, VARIANT& variant);
+
+    template <class RET, size_t INDEX>
+    static RET& getImpl(bsl::in_place_index_t<INDEX>, const VARIANT& variant);
+
+    static bool lessThanImpl(const VARIANT&, const VARIANT&);
+
+    static bool equalImpl(const VARIANT&, const VARIANT&);
+};
+
+#ifdef BSL_VARIANT_FULL_IMPLEMENTATION
+
+                         // ==========================
+                         // struct Variant_ArrayHelper
+                         // ==========================
+template <class TYPE>
+struct Variant_ArrayHelper {
+    // This metafunction is used to satisfy the standard requirement on
+    // converting constructors which is expressed in terms of
+    // 'x[] = {std::forward<TYPE>(t)};'
+    TYPE x[1];
+};
+
+/* TODO : remove before release ?
+ * This checks whether P1957R1 is implemented, in case a flag needs to be
+ * introduced for narrowing conversion. Currently, the check for narrowing
+ * conversion has a special case for bool destination type which should cover
+ * the compilers which do not yet have P1957R1.
+ * */
+template <class TYPE, class = void>
+struct Variant_CheckForP1957R1 : bsl::true_type {
+};
+
+template <class TYPE>
+struct Variant_CheckForP1957R1<
+    TYPE,
+    bsl::void_t<decltype(Variant_ArrayHelper<TYPE>{{"bc"}})> >
+: bsl::false_type {
+};
+//static_assert(Variant_CheckForP1957R1<bool>::value);
+
+                  // =======================================
+                  // struct Variant_ConvertsWithoutNarrowing
+                  // =======================================
+template <class DEST, class SOURCE, typename = void>
+struct Variant_ConvertsWithoutNarrowing : bsl::false_type {
+};
+
+template <class DEST, class SOURCE>
+struct Variant_ConvertsWithoutNarrowing<
+    DEST,
+    SOURCE,
+    bsl::void_t<decltype(
+        Variant_ArrayHelper<DEST>{{std::declval<SOURCE>()}})> >
+: bsl::conditional<
+      !Variant_CheckForP1957R1<bool>::value &&
+          bsl::is_same<bool, typename bsl::remove_cvref<DEST>::type>::value &&
+          !bsl::is_same<bool, typename bsl::remove_cvref<SOURCE>::type>::value,
+      bsl::false_type,
+      bsl::true_type>::type {
```

Comment:
The use of `bsl::conditional` here seems unnecessary because it's not being used to prevent any (potentially expensive or ill-formed) instantiations in the not-taken branch. This should just be
```suggestion
: bsl::integral_constant<
      bool,
      Variant_CheckForP1957R1<bool>::value ||
          !bsl::is_same<bool, typename bsl::remove_cvref<DEST>::type>::value ||
          bsl::is_same<bool,
                       typename bsl::remove_cvref<SOURCE>::type>::value> {
```
(At least, that's the formatting that bde-format spits out for me.)

Reply:
not sure why the conditional, possibly a left over from some previous version. Changing to `integral_constant`.
May i suggest we do the flipped condition, i.e.
``` C++
: bsl::integral_constant<
      bool,
      !(!Variant_CheckForP1957R1<bool>::value &&
        bsl::is_same<bool, typename bsl::remove_cvref<DEST>::type>::value &&
        !bsl::is_same<bool,
                      typename bsl::remove_cvref<SOURCE>::type>::value)> {
  ```            
   I think it makes it slightly more humanly parsable (if P1957 isn't implemented, the DEST is bool, and the SOURCE isn't bool). WDYT ?

Reply:
I think having it flipped is fine

Reply:
done